binary_gentoo/internal/cli/tree_diff.py

```diff
@@ -3,12 +3,72 @@
 
 import filecmp
 import os
+import re
 import sys
 from argparse import ArgumentParser
+from typing import Iterable, Set
 
-from ..reporter import exception_reporting
+from ..reporter import announce_and_check_output, exception_reporting
 from ._parser import add_version_argument_to
 
+_keywords_pattern = re.compile('KEYWORDS="(?P<keywords>[^"]*)"')
+
+
+def _get_relevant_keywords_set_for(ebuild_content: str,
+                                   accept_keywords: Iterable[str]) -> Set[str]:
+    match = _keywords_pattern.search(ebuild_content)
+    if match is None:
+        # if the KEYWORDS variable is not found in the ebuild,
+        # we will assume the ebuild needs to be listed
+        ebuild_keywords = accept_keywords
+    else:
+        ebuild_keywords = match.group('keywords')
+        ebuild_keywords = [kw for kw in ebuild_keywords.split(" ") if kw]
+
+    # replace special keywords by the relevant keywords from the ebuild
+    if '**' in accept_keywords:
+        accept_keywords = ebuild_keywords
+    elif '*' in accept_keywords:
+        accept_keywords = [kw for kw in ebuild_keywords if not kw.startswith('~')]
+    elif '~*' in accept_keywords:
+        accept_keywords = [kw for kw in ebuild_keywords if kw.startswith('~')]
```

Comment:
Indeed, I took some wrong assumptions here. I will correct.

About the typing - indeed, I'm going to explicitly use Set since these adaptations will require a Set as input.

I'll look into some reorganization. That will be necessary based on your other comments anyway.